modules/arcore/arcore_interface.cpp

```diff
@@ -0,0 +1,696 @@
+
+// TODO remove these temporary includes once we add our GL_TEXTURE_EXTERNAL_OES support to our drivers
+#include "platform_config.h"
+#ifndef GLES3_INCLUDE_H
+#include <GLES3/gl3.h>
+#else
+#include GLES3_INCLUDE_H
+#endif
+
+/**
+	@author Bastiaan Olij <mux213@gmail.com>, Robert Hofstra <robert.hofstra@knowlogy.nl>
+	ARCore interface between Android and Godot
+**/
+
+StringName ARCoreInterface::get_name() const {
+	return "ARCore";
+}
+
+int ARCoreInterface::get_capabilities() const {
+	return ARVRInterface::ARVR_MONO + ARVRInterface::ARVR_AR;
+}
+
+int ARCoreInterface::get_camera_feed_id() {
+	if (feed.is_valid()) {
+		return feed->get_id();
+	} else {
+		return 0;
+	}
+}
+
+void ARCoreInterface::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("_resume"), &ARCoreInterface::_resume);
+}
+
+bool ARCoreInterface::is_initialized() const {
+	// if we're in the process of initialising we treat this as initialised...
+	return (init_status != NOT_INITIALISED) && (init_status != INITIALISE_FAILED);
+}
+
+void ARCoreInterface::_resume() {
+	if (ar_session == NULL) {
+		// Most of this code is based on the hello ar sample application from Google. Most remarks are
+		// theirs and we'll need to do something with the suggestions made within.
+		ArInstallStatus install_status;
+		OS_Android *os_android = (OS_Android *)OS::get_singleton();
+
+		print_line("Godot ARCore: Checking camera");
+
+		// Check our camera privilege
+		// TODO: Note that if we asked for permission already we should just
+		// check if we have permission and fail if we don't, user said no...
+		if (!os_android->request_permission("CAMERA")) {
+			init_status = REQUESTED_CAMERA;
+			return;
+		}
+
+		print_line("Godot ARCore: Getting environment");
+
+		// get some android things
+		JNIEnv *env = ThreadAndroid::get_env();
+
+		godot_java = os_android->get_godot_java();
+		jobject activity = godot_java->get_activity();
+		if (activity == NULL) {
+			print_line("Godot ARCore: Couldn't get activity");
+			init_status = INITIALISE_FAILED; // don't try again.
+			return;
+		}
+		jobject context = godot_java->get_application_context();
+		if (context == NULL) {
+			print_line("Godot ARCore: Couldn't get context");
+			init_status = INITIALISE_FAILED; // don't try again.
+			return;
+		}
+
+		// If install was not yet requested, that means that we are resuming the
+		// activity first time because of explicit user interaction (such as
+		// launching the application)
+		bool user_requested_install = init_status == WAITING_ON_INSTALL ? 0 : 1;
+
+		print_line("Godot ARCore: Request install.");
+
+		// === ATTENTION!  ATTENTION!  ATTENTION! ===
+		// This method can and will fail in user-facing situations.  Your
+		// application must handle these cases at least somewhat gracefully.  See
+		// HelloAR Java sample code for reasonable behavior.
+		ArStatus status = ArCoreApk_requestInstall(env, activity, user_requested_install, &install_status);
+		if (status != AR_SUCCESS) {
```

Comment:
@BastiaanOlij Another advantage of leveraging the XR Mode interface. We should have some ARCore specific init logic in the Godot.java and GodotView.java which among other things checks that ARCore is installed, and if not, handles the error gracefully.
This will ensure that by the time we get to the native code, all the ARCore install and setup is completed, and we don't need to worry about error cases here.

I'll try to send you a PR to illustrate what I'm describing.

Reply:
Not sure if that has a whole lot of benefit but I guess it makes sense to have a simple place where permissions are all checked.

Reply:
One key advantage to this approach is that in some scenarios, the ARCore install and setup on the device is not fully complete, and the ARCore apis requires the app to fire an intent in order to resolve that. 
This is very hard to do from the native code (e.g: require a lot of plumbing), but very simple to do from the java Android activity layer.